lib/topic_view.rb

```diff
@@ -604,8 +604,9 @@ def setup_filtered_posts
 
     if SiteSetting.ignore_user_enabled
       @filtered_posts = @filtered_posts.where.not("user_id IN (?) AND id <> ?",
-                                                  IgnoredUser.where(user_id: @user.id).select(:ignored_user_id),
+                                                  IgnoredUser.where(user_id: @user&.id).select(:ignored_user_id),
```

Comment:
A few thoughts here:

1. `pluck` is much more efficient than `select` if you want an array of values returned. `select` will create object instances.

2. If the ignored user query returns nothing, we don't need to execute the `where` query. Most users will not have any ignored users so you can often save a query by first retrieving the ignored user ids into a variable, and then only executing the filter query if that array is not empty.

3. At some point we should benchmark the cost of querying the ignored users versus per topic view storing them in a cache and see if we can win some performance that way.

Reply:
(also sorry I know none of those thoughts have to do with this FIX, but I just noticed them!)

Reply:
> 1. `pluck` is much more efficient than `select`, ....

Makes sense 👍 

> 2. If the ignored user query returns nothing, .....

That makes sense, it is quick to change but since this is experimental and I aim to do some changes in this path again, would it be okay to revisit later?

> 3.  At some point we should benchmark

👍

Reply:
> * `pluck` is much more efficient than `select` if you want an array of values returned. `select` will create object instances.

Except in this case, the `select` won't return an array. It'll be directly transformed into SQL 😉 

At some point we should benchmark the differences between

- Doing a small request with a `pluck` and then adding the `where` if there are ignored users

``` ruby
ignored_user_ids = IgnoredUser.where(user_id: @user&.id).pluck(:ignored_user_id)

if ignored_user_ids.present?
  @filtered_posts = @filtered_posts.where.not("user_id IN (?) AND id <> ?", ignored_user_ids, first_post_id)
end
```

- Using `select` and always adding the `where` clause

``` ruby
ignored_users = IgnoredUser.where(user_id: @user&.id).select(:ignored_user_id)
@filtered_posts = @filtered_posts.where.not("user_id IN (?) AND id <> ?", ignored_users, first_post_id)
```

@SamSaffron any thoughts about this?

Reply:
Although, my point is kinda moot if we need to know if there are ignored users to set the `@contains_gap` to `true`.

Reply:
> Except in this case, the select won't return an array. It'll be directly transformed into SQL 😉

`select` will result in an active relation, but `pluck` will directly return an array of IDs.

Reply:
Yes, but that's way less memory that loading the `IgnoredUser` records in memory ;) (which happens if you iterate over the results of the `select`)

Reply:
I am happy to go with either approaches. I am unsure, in this PR's context (besides future bench marking) is there any changes that we need to do?

Reply:
I think it's fine as is for now. We can revisit later.